### PR TITLE
reland "fix the widget span layout when text scale factor != 1" and h…

### PR DESCRIPTION
**Pull request by chunhtai**
…ide it behind a flag

## Description

This PR relands https://github.com/flutter/flutter/pull/59711 and hide the entire fix behind a flag so that i can do soft transition in our internal tests.

## Related Issues

https://github.com/flutter/flutter/issues/59316

## Tests

I added the following tests:

see files

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by justinmc — APPROVED** [review]
LGTM

Was the failure that caused the revert just visual diff tests that captured the incorrect layout?

**Comment by chunhtai** [comment]
> LGTM
> 
> Was the failure that caused the revert just visual diff tests that captured the incorrect layout?

Mostly yes, but there are also some tests that creates columns that do not have enough height and causes overflow.

File: packages/flutter/lib/src/rendering/paragraph.dart [hunk 1]
```diff
@@ -79,6 +79,13 @@ class RenderParagraph extends RenderBox
     StrutStyle strutStyle,
     TextWidthBasis textWidthBasis = TextWidthBasis.parent,
     ui.TextHeightBehavior textHeightBehavior,
+    @Deprecated(
+      'This parameter is a temporary flag to migrate the internal tests and '
+      'should not be used in other context. For more details, please check '
```
Comment (HansMuller) [inline packages/flutter/lib/src/rendering/paragraph.dart:84]: context => contexts.

Here and in the other copies of this comment

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by flutter-github-sync** [comment]
Google testing passed!

**Comment by flutter-github-sync** [comment]
Google testing passed!

**Pull request merged**